ethdb/privateapi/ethbackend.go

```diff
@@ -678,6 +678,78 @@ func (s *EthBackendServer) EngineForkChoiceUpdated(ctx context.Context, req *rem
 	}, nil
 }
 
+func (s *EthBackendServer) EngineGetPayloadBodiesByHashV1(ctx context.Context, request *remote.EngineGetPayloadBodiesByHashV1Request) (*remote.EngineGetPayloadBodiesV1Response, error) {
+	tx, err := s.db.BeginRo(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	bodies := make([]*types2.ExecutionPayloadBodyV1, len(request.Hashes))
+
+	for hashIdx, hash := range request.Hashes {
+		h := gointerfaces.ConvertH256ToHash(hash)
+		block, err := rawdb.ReadBlockByHash(tx, h)
+		if err != nil {
+			return nil, err
+		}
+
+		bodies[hashIdx] = extractPayloadBodyFromBlock(block)
+	}
+
+	return &remote.EngineGetPayloadBodiesV1Response{Bodies: bodies}, nil
+}
+
+func (s *EthBackendServer) EngineGetPayloadBodiesByRangeV1(ctx context.Context, request *remote.EngineGetPayloadBodiesByRangeV1Request) (*remote.EngineGetPayloadBodiesV1Response, error) {
+	tx, err := s.db.BeginRo(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	bodies := make([]*types2.ExecutionPayloadBodyV1, request.Count)
+
+	var i uint64
+	for i = 0; i < request.Count; i++ {
+		block, err := rawdb.ReadBlockByNumber(tx, request.Start+i)
+		if err != nil {
+			return nil, err
+		}
+		bodies[i] = extractPayloadBodyFromBlock(block)
+	}
+
+	return &remote.EngineGetPayloadBodiesV1Response{Bodies: bodies}, nil
+}
+
+func extractPayloadBodyFromBlock(block *types.Block) *types2.ExecutionPayloadBodyV1 {
+	if block == nil {
+		return nil
+	}
+
+	txs := block.Transactions()
+	bdTxs := make([][]byte, len(txs))
+	for idx, tx := range txs {
+		bdTxs[idx] = tx.GetData()
```

Comment:
Should be `MarshalBinary` (for typed transactions their [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) marshalling not additionally wrapped into RLP – see `TestEIP2718TransactionEncode`) rather than `GetData`.

Reply:
Pushed some changes here, thanks for the heads up, I wouldn't have picked up on that one!  To keep things straight forward for now put a nil into the slice for errors to keep ordering as expected.  Might be better to return an error at this point?

Reply:
It's better to return an error.

Reply:
OK brill, code updated to bomb early